### DISCUSSION TITLE
Does frb work with safari?

### DISCUSSION THREAD

**Discussion by uditrugman** [fzyzcjy/flutter_rust_bridge#1611] [Q&A]
I'm trying to run flutter_via_create from frb_examples in safari and I getting the below error:

[Error] 
	ae_ (main.dart.js:6353)
	ae0 (main.dart.js:6437)
	$0 (main.dart.js:30611)
	ajl (main.dart.js:5802)
	ajx (main.dart.js:5804)
	$1 (main.dart.js:30044)

is safari support with frb?

**Comment by fzyzcjy** [comment]
Hi, could you please show more errors? As a first guess without more error logs, I guess Safari may not support some Rust WASM features that Rust code needs. flutter_rust_bridge itself does not require anything (since it merely generates some normal code), but Rust (on Web) may require some.

**Comment by uditrugman** [comment]
These are the only logs.

[Debug] Installing/Activating first service worker. (flutter.js, line 172)
[Debug] Activated new service worker. (flutter.js, line 205)
[Debug] Injecting <script> tag. Using callback. (flutter.js, line 298)
[Debug] Flutter Web Bootstrap: Programmatic. (main.dart.js, line 7856)
[Error] 
	ae_ (main.dart.js:6353)
	ae0 (main.dart.js:6437)
	$0 (main.dart.js:30611)
	ajl (main.dart.js:5802)
	ajx (main.dart.js:5804)
	$1 (main.dart.js:30044)



i tested some other rust examples and they worked. here is one example:
https://wasmbyexample.dev/examples/hello-world/hello-world.assemblyscript.en-us

**Reply by fzyzcjy** [reply]
Hmm, looks like the stack trace is obfuscated. Could you please build it in non-release mode?

**Comment by uditrugman** [comment]
I did not run with obfuscation, unless it's the default mode.
is there a way to run in release mode without obfuscation?

below are the logs when a run it in debug mode, but now, i believe the errors are because safari is looking for the devtools, so it might be unreleated to the wasm issue.

```
[Error] Error from backend:

Error: Bad state: Not connected to an application.

Stack Trace:
#0      DevHandler._handleConnection.<anonymous closure> (package:dwds/src/handlers/dev_handler.dart:271:13)
#1      _rootRunUnary (dart:async/zone.dart:1407:47)
#2      _CustomZone.runUnary (dart:async/zone.dart:1308:19)
#3      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1217:7)
#4      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
#5      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#6      _ForwardingStreamSubscription._add (dart:async/stream_pipe.dart:123:11)
#7      _MapStream._handleData (dart:async/stream_pipe.dart:218:10)
#8      _ForwardingStreamSubscription._handleData (dart:async/stream_pipe.dart:153:13)
#9      _rootRunUnary (dart:async/zone.dart:1415:13)
#10     _CustomZone.runUnary (dart:async/zone.dart:1308:19)
#11     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1217:7)
#12     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
#13     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#14     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:784:19)
#15     _StreamController._add (dart:async/stream_controller.dart:658:7)
#16     _StreamController.add (dart:async/stream_controller.dart:606:5)
#17     new WebSocketImpl.fromSocket.<anonymous closure> (package:web_socket_channel/src/copy/web_socket_impl.dart:730:21)
#18     _rootRunUnary (dart:async/zone.dart:1407:47)
#19     _CustomZone.runUnary (dart:async/zone.dart:1308:19)
#20     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1217:7)
#21     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
#22     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#23     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:63:11)
#24     _EventSinkWrapper.add (dart:async/stream_transformers.dart:13:11)
#25     _WebSocketProtocolTransformer._messageFrameEnd (package:web_socket_channel/src/copy/web_socket_impl.dart:319:23)
#26     _WebSocketProtocolTransformer.add (package:web_socket_channel/src/copy/web_socket_impl.dart:218:46)
#27     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:111:24)
#28     _rootRunUnary (dart:async/zone.dart:1407:47)
#29     _CustomZone.runUnary (dart:async/zone.dart:1308:19)
#30     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1217:7)
#31     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
#32     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#33     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:784:19)
#34     _StreamController._add (dart:async/stream_controller.dart:658:7)
#35     _StreamController.add (dart:async/stream_controller.dart:606:5)
#36     _Socket._onData (dart:io-patch/socket_patch.dart:2447:41)
#37     _rootRunUnary (dart:async/zone.dart:1415:13)
#38     _CustomZone.runUnary (dart:async/zone.dart:1308:19)
#39     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1217:7)
#40     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
#41     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#42     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:784:19)
#43     _StreamController._add (dart:async/stream_controller.dart:658:7)
#44     _StreamController.add (dart:async/stream_controller.dart:606:5)
#45     new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1936:33)
#46     _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:1379:14)
#47     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#48     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)
#49     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#50     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:185:5)

	(anonymous function) (client.js:24977)
	(anonymous function) (client.js:3589)
	call$2 (client.js:11698)
	_asyncStartSync (client.js:3553)
	$call$body$main__closure (client.js:24985)
	call$1 (client.js:24911)
	_rootRunUnary (client.js:3948)
	runUnary$2$2 (client.js:13095)
	runUnaryGuarded$1$2 (client.js:13042)
	_sendData$1 (client.js:12635)
	_add$1 (client.js:12581)
	_add$1 (client.js:12916)
	_handleData$2 (client.js:12979)
	_handleData$1 (client.js:12942)
	(anonymous function) (client.js:1348)
	_rootRunUnary (client.js:3948)
	runUnary$2$2 (client.js:13095)
	runUnaryGuarded$1$2 (client.js:13042)
	_sendData$1 (client.js:12635)
	_add$1 (client.js:12581)
	_sendData$1 (client.js:12427)
	add$1 (client.js:12306)
	(anonymous function) (client.js:1406)
	_rootRunUnary (client.js:3948)
	runUnary$2$2 (client.js:13095)
	runUnaryGuarded$1$2 (client.js:13042)
	_sendData$1 (client.js:12635)
	_add$1 (client.js:12581)
	_sendData$1 (client.js:12427)
	add$1 (client.js:12306)
	add$1 (client.js:24411)
	call$1 (client.js:24701)
	call$1 (client.js:18104)
	_rootRunUnary (client.js:3952)
	runUnary$2$2 (client.js:13095)
	runUnaryGuarded$1$2 (client.js:13042)
	call$1 (client.js:13232)
	invokeClosure (client.js:1238)
	(anonymous function) (client.js:1257)

**Reply by fzyzcjy** [reply]
I will have a look at the problem, hopefully in a few days (though I suspect it is a limitation of Rust, and flutter_rust_bridge can do nothing to solve it).

**Comment by uditrugman** [comment]
could it be related to "init" being async?
maybe this causes frb to create threads which are not available on Safari.

**Reply by fzyzcjy** [reply]
Looks reasonable! Then one way may be disable some frb features on safari. 

But I guess the first step is to get a non-obfuscated stacktrace, then it will be much easier to see what is going on.

**Comment by uditrugman** [comment]
how can i disable frb features and which ones would you suggest to disable?

and, I'm not sure how to get non-obfuscated stacktrace. are you sure they are obfuscated and not just failing on some code that does not have symbols?

**Reply by fzyzcjy** [reply]
> how can i disable frb features and which ones would you suggest to disable?

I mean make a PR to do so, since I (and other people) did not thought about this safari issue before imho, so frb does not have a feature for that yet.

> are you sure they are obfuscated and not just failing on some code that does not have symbols?

Good question, it is also possible!

**Comment by uditrugman** [comment]
i added some logging and it appears that crossOriginIsolated is false which causes an exception to be thrown just before loading the WASM module.

according to the comment (below), it appears that frb requires that buffers can be shared between workers.
I'd guess that Safari does not allow that.

Any suggestion on what is needed to make frb work without shared buffers between workers?
I assume that it also means that frb should also support running without threads and i've read somewhere that it already supports that mode.

```
/// Whether the web platform has been isolated by COOP and COEP headers,
/// and is capable of sharing buffers between workers.
///
/// Note: not available on all browsers, in which case it will return null.
@JS()
external bool? get crossOriginIsolated;

**Reply by fzyzcjy** [reply]
Looks reasonable! If safari does not allow it (maybe check this first), we will need to forbid these features, e.g. only allow execution on the main thread, or use other mechanisms to send buffers.

**Reply by fzyzcjy** [reply]
Ok I see: https://cjycode.com/flutter_rust_bridge/manual/miscellaneous/wasm-limitations

> Safari cannot spawn nested Workers. A workaround is to build two variants of the library, one with multithreading and one without, and serve Safari users the single-threaded variant. For a more general solution, check out [wasm-feature-detect](https://github.com/GoogleChromeLabs/wasm-feature-detect) or [nested-worker](https://github.com/johanholmerin/nested-worker).

Also /cc @Desdaemon who PRed the great web feature, in case he has some more insights

**Reply by fzyzcjy** [reply]
Btw, you can reply below me (instead of creating a new answer), then IIRC it will be more ordered even when votes change

**Reply by uditrugman** [reply]
ok,
so, in order to test, i made the init_app function synchronous, then commented out the  "_ensureCrossOriginIsolated" function.

and now the it's running in Safari :)

I guess the next step is to add a feature that disable threading, right?

**Reply by fzyzcjy** [reply]
Good job! I think so.

**Reply by fzyzcjy** [reply]
Or, alternatively, if you never use the thread-pool-based function (`fn`), or always use `async fn` (this *is* async, but runs on main thread on web IIRC), I guess this should be OK?

**Reply by fzyzcjy** [reply]
Btw, another way is to change thread pool implementation https://github.com/fzyzcjy/flutter_rust_bridge/blob/54d60706ecc2490dc6a57032b36da80196bc51ed/frb_rust/src/handler/implementation/executor.rs#L64, by using a single-theead "pseudo" thread pool on it.

I will check this issue in more details later (as mentioned earlier)

**Reply by fzyzcjy** [reply]
I will continue discussion on https://github.com/fzyzcjy/flutter_rust_bridge/issues/1618, since this looks like a reasonable enhancement :)
